### PR TITLE
feat: make risk assessment a context property

### PR DESCRIPTION
**Pull request by remisultan**
## :id: Reference related issue. 

https://github.com/gravitee-io/issues/issues/7555
https://github.com/gravitee-io/issues/issues/7657

## :pencil2: A description of the changes proposed in the pull request

This PR brings the RiskAssessment into AccessManagement
It also introduces UserActivity in order to collect data to apply risk assessment when activated

The endUser can consent at login to have some information stored and processed to be used for RiskAssessment
You can activate it in the gravitee.yml (both management and gateway)

```yml
 user:
   activity:
      enabled: false|true # default is false
      anon:
         algorithm: SHA256|SHA512|NONE #default SHA256
         salt: some-salt # default is null meaning the key generated will change every time
      retention:
         time: 3
         unit: MONTHS
      geolocation:
         variation:
            latitude: 0.07 # default to have a geolocation randomised, 0 will give the exact position
            longitude: 0.07 # default to have a geolocation randomised, 0 will give the exact position
```

The `retention.unit` option is based on `java.time.temporal.ChronoUnit`

## :memo: Test scenarios 

Put the above in your gravitee.yml and set enabled to true

You will need:
      - https://github.com/gravitee-io/gravitee-risk-assessment/pull/2
      - https://github.com/gravitee-io/gravitee-service-geoip

built and zipped in the `plugins/`.

Login at least twice to have data configured.

On the login page + webauthn page, you can use
- `uc_geoip`
- `uc_ua`
to validate consent to use users' data from an input tag (with a checkbox or with a hidden input with the value `on|off`)

## :computer: Add screenshots for UI

image

image

image

### CODE REVIEW

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        return userActivityRepository.findByDomainAndTypeAndKeyAndLimit(domain, type, buildKey(userId), limit);
+    }
+
+    public Completable save(
+            String domain,
+            String userId,
+            UserActivity.Type type,
+            Map<String, Object> data) {
+        final Date createdAt = new Date();
+        return Single.defer(() -> Single.just(new UserActivity()
```
Comment (leleueri) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java:86]: Why are we using a defer here to create a Bean ?

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    public Completable deleteByDomain(String domain) {
+        return userActivityRepository.deleteByDomain(domain).doOnError(err ->
+                LOGGER.error("An unexpected error has occurred while deleting userActivity '{}'", err.getMessage(), err)
+        );
+    }
+
+    private String buildKey(String userId) {
+        return computeHash(configuration.getAlgorithmKey(), userId, configuration.getSalt());
```
Comment (leleueri) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java:119]: is it safe to have the same salt value for each user?
Wondering if a random value generated with a seed coming from the userId is possible :thinking:

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private Double buildLatitude(Map<String, Object> data) {
+        return computeCoordinate(data, LATITUDE_KEY, configuration.getLatitudeVariation(), 90);
```
Comment (leleueri) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java:123]: What does 90 mean ? Maybe a Constant to name it ?

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private Double buildLongitude(Map<String, Object> data) {
+        return computeCoordinate(data, LONGITUDE_KEY, configuration.getLongitudeVariation(), 180);
```
Comment (leleueri) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java:127]: What does 180 mean ? Maybe a Constant to name it ?

File: gravitee-am-service/src/test/java/io/gravitee/am/service/impl/user/activity/UserActivityServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        testObserver.assertComplete();
+        testObserver.assertNoErrors();
+    }
+
+    @Test
+    public void must_throw_error_when_delete_by_domain_and_key() {
+        doReturn(Completable.error(new IllegalArgumentException("An Error"))).when(userActivityRepository).deleteByDomainAndKey(anyString(), anyString());;
+        final TestObserver<Void> testObserver = userActivityService.deleteByDomainAndUser("domain", "user-id").test();
+
+        testObserver.assertError(IllegalArgumentException.class);
```
Comment (leleueri) [inline gravitee-am-service/src/test/java/io/gravitee/am/service/impl/user/activity/UserActivityServiceTest.java:120]: missing awaitTerminalEvent that may lead to unstable test

File: gravitee-am-service/src/test/java/io/gravitee/am/service/impl/user/activity/UserActivityServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    @Test
+    public void must_throw_error_when_delete_by_domain() {
+        doReturn(Completable.error(new IllegalArgumentException("An Error"))).when(userActivityRepository).deleteByDomain(anyString());
+        final TestObserver<Void> testObserver = userActivityService.deleteByDomain("domain").test();
+
+        testObserver.assertError(IllegalArgumentException.class);
```
Comment (leleueri) [inline gravitee-am-service/src/test/java/io/gravitee/am/service/impl/user/activity/UserActivityServiceTest.java:128]: missing awaitTerminalEvent that may lead to unstable test

**Review by leleueri — CHANGES_REQUESTED** [review]

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java [hunk 7]
```diff
@@ -62,40 +55,21 @@
...
-import io.gravitee.am.gateway.handler.root.resources.handler.user.password.ResetPasswordOneTimeTokenHandler;
-import io.gravitee.am.gateway.handler.root.resources.handler.user.password.ResetPasswordRequestParseHandler;
-import io.gravitee.am.gateway.handler.root.resources.handler.user.password.ResetPasswordSubmissionRequestParseHandler;
-import io.gravitee.am.gateway.handler.root.resources.handler.user.register.RegisterAccessHandler;
-import io.gravitee.am.gateway.handler.root.resources.handler.user.register.RegisterConfirmationRequestParseHandler;
-import io.gravitee.am.gateway.handler.root.resources.handler.user.register.RegisterConfirmationSubmissionRequestParseHandler;
-import io.gravitee.am.gateway.handler.root.resources.handler.user.register.RegisterFailureHandler;
-import io.gravitee.am.gateway.handler.root.resources.handler.user.register.RegisterProcessHandler;
-import io.gravitee.am.gateway.handler.root.resources.handler.user.register.RegisterSubmissionRequestParseHandler;
+import io.gravitee.am.gateway.handler.root.resources.handler.user.password.*;
```
[the remark this answers is not in the document]
Comment (remisultan) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java:65]: I agree.

My IDE is being naughty :D

File: gravitee-am-common/src/main/java/io/gravitee/am/common/utils/ConstantKeys.java [hunk 8]
```diff
@@ -149,4 +149,11 @@ public interface ConstantKeys {
     String HTTP_SSL_ALIASES_ENDPOINTS_CIBA = "backchannel_authentication_endpoint";
 
     String CIBA_AUTH_REQUEST_KEY = "ciba_authentication_request";
+
+    //User Activity
+    String USER_CONSENT_IP_LOCATION = "user_consent_ip_location";
+    String USER_CONSENT_USER_AGENT = "user_consent_user_agent";
+    String USER_ACTIVITY_RETENTION_TIME = "user_activity_retention_time";
+    String RISK_ASSESSMENT_KEY = "risk_assessment";
```
[the remark this answers is not in the document]
Comment (remisultan) [reply gravitee-am-common/src/main/java/io/gravitee/am/common/utils/ConstantKeys.java:157]: What do you think of:
- `uc_geoip`
- `uc_ua`
- `ua_rti`

File: gravitee-am-model/src/main/java/io/gravitee/am/model/UserActivity.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+package io.gravitee.am.model;
+
+import java.util.Date;
+
+/**
+ * @author Rémi SULTAN (remi.sultan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+// TODO: Is this the name we want ?
```
[the remark this answers is not in the document]
Comment (remisultan) [reply gravitee-am-model/src/main/java/io/gravitee/am/model/UserActivity.java:25]: To me this is going to be useful the more we dive into UEBA.
Today we're using it for RiskAssessment but UEBA goes beyond just assessing risk. 

It gives information on user behaviour and activity

File: gravitee-am-repository/gravitee-am-repository-jdbc/src/main/resources/liquibase/changelogs/v3_18_0/schema.yml [hunk 10]
```diff
@@ -0,0 +1,52 @@
+databaseChangeLog:
+  - changeSet:
+      id: 3.18.0
```
[the remark this answers is not in the document]
Comment (remisultan) [reply gravitee-am-repository/gravitee-am-repository-jdbc/src/main/resources/liquibase/changelogs/v3_18_0/schema.yml:3]: Ok for me

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java (hunk 1 above)
Comment (remisultan) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java:86]: We're building the hashed key of a user so we might want to defer in order to compute the hashed key when needed

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java (hunk 2 above)
Comment (remisultan) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java:119]: Let's disscuss about it

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java (hunk 3 above)
Comment (remisultan) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java:123]: We'll change this

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/webauthn/WebAuthnLoginEndpoint.java [hunk 11]
```diff
@@ -152,20 +160,27 @@ private void getCredentials(RoutingContext ctx) {
...
                                 var deviceType = webauthnLogin.getString(DEVICE_TYPE);
-                                checkIfDeviceExists(ctx, client, h.result().getId(), deviceId, deviceType, rememberDeviceSettings)
-                                        .doFinally(() -> buildResponse(ctx, getAssertion))
-                                        .subscribe();
-                            } else {
-                                buildResponse(ctx, getAssertion);
+                                completable = checkIfDeviceExists(ctx, client, user.getId(), deviceId, deviceType, rememberDeviceSettings);
                             }
+                            if (userActivityService.canSaveUserActivity()) {
+                                completable = getUserActivityIfNecessary(ctx, user, completable);
```
[the remark this answers is not in the document]
Comment (remisultan) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/webauthn/WebAuthnLoginEndpoint.java]: I implemented that way to avoir `if else` trivia:

- Only remember device active
- Only user activity active
- both active

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/risk/RiskAssessmentHandler.java [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        } else {
+            routingContext.next();
+        }
+    }
+
+    private void computeRiskAssessment(RoutingContext routingContext, Client client, io.gravitee.am.model.User user, RiskAssessmentSettings riskAssessment) {
+        var assessmentMessage = Single.just(new AssessmentMessage().setSettings(riskAssessment).setData(new AssessmentData()));
+        assessmentMessage = buildDeviceAssessmentMessage(routingContext, riskAssessment, assessmentMessage, client, user);
+        assessmentMessage = buildIpReputationMessage(routingContext, riskAssessment, assessmentMessage);
+        assessmentMessage = buildGeoVelocityMessage(client, user, riskAssessment, assessmentMessage);
```
[the remark this answers is not in the document]
Comment (remisultan) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/handler/risk/RiskAssessmentHandler.java:101]: That'd make more sense indeed

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/webauthn/WebAuthnLoginEndpoint.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (leleueri) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/webauthn/WebAuthnLoginEndpoint.java]:
```suggestion
                        Completable completable = Completable.complete();
                        if (rememberDeviceSettings.isActive()) {
                            var deviceId = webauthnLogin.getString(DEVICE_ID);
                            var deviceType = webauthnLogin.getString(DEVICE_TYPE);
                            completable = completable.andThen(checkIfDeviceExists(ctx, client, user.getId(), deviceId, deviceType, rememberDeviceSettings));
                        }
                        if (userActivityService.canSaveUserActivity()) {
                            completable = completable.andThen(getUserActivityIfNecessary(ctx, user, completable));
                        }
                        completable
                                .doOnError(err -> logger.error("An unexpected error has occurred: {}", err.getMessage(), err))
                                .doFinally(() -> buildResponse(ctx, getAssertion))
                                .subscribe();
```

File: gravitee-am-model/src/main/java/io/gravitee/am/model/UserActivity.java [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+    private Type userActivityType;
+    private String userActivityKey;
+
+    private Double latitude;
+    private Double longitude;
+    private String userAgent;
+    private Integer loginAttempts;
+
+    private Date createdAt;
+    private Date updatedAt;
```
[the remark this answers is not in the document]
Comment (leleueri) [reply gravitee-am-model/src/main/java/io/gravitee/am/model/UserActivity.java:46]: Don't know of we have to keep it or not to be honest, do you think a User Activity can be updated in a near future ?

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/webauthn/WebAuthnLoginEndpoint.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (remisultan) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/webauthn/WebAuthnLoginEndpoint.java]: Fine 👍

**Comment by remisultan** [comment]
closed due to https://github.com/gravitee-io/gravitee-access-management/pull/1912

**Pull request closed without merging**